homeassistant/components/template/button.py

```diff
@@ -0,0 +1,157 @@
+"""Support for buttons which integrates with other components."""
+from __future__ import annotations
+
+import contextlib
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.button import (
+    DEVICE_CLASSES_SCHEMA,
+    DOMAIN as BUTTON_DOMAIN,
+    ButtonDeviceClass,
+    ButtonEntity,
+)
+from homeassistant.components.template import TriggerUpdateCoordinator
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_ICON, CONF_NAME, CONF_UNIQUE_ID
+from homeassistant.core import Config, HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.script import Script
+from homeassistant.helpers.template import Template, TemplateError
+
+from .const import CONF_AVAILABILITY, DOMAIN
+from .template_entity import TemplateEntity
+from .trigger_entity import TriggerEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_PRESS = "press"
+
+DEFAULT_NAME = "Template Button"
+DEFAULT_OPTIMISTIC = False
+
+BUTTON_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.template,
+        vol.Required(CONF_PRESS): cv.SCRIPT_SCHEMA,
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_AVAILABILITY): cv.template,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
+        vol.Optional(CONF_ICON): cv.template,
+    }
+)
+
+
+async def _async_create_entities(
+    hass: HomeAssistant, definitions: list[dict[str, Any]], unique_id_prefix: str | None
+) -> list[TemplateButtonEntity]:
+    """Create the Template button."""
+    entities = []
+    for definition in definitions:
+        unique_id = definition.get(CONF_UNIQUE_ID)
+        if unique_id and unique_id_prefix:
+            unique_id = f"{unique_id_prefix}-{unique_id}"
+        entities.append(
+            TemplateButtonEntity(
+                hass,
+                definition[CONF_NAME],
+                definition.get(CONF_AVAILABILITY),
+                definition[CONF_PRESS],
+                definition.get(CONF_DEVICE_CLASS),
+                unique_id,
+                definition.get(CONF_ICON),
+            )
+        )
+    return entities
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: Config,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: dict[str, Any] | None = None,
+) -> None:
+    """Set up the template button."""
+    if discovery_info is None:
+        _LOGGER.warning(
```

Comment:
When does this happen? Please cover in a test.

Reply:
this happens when someone uses `button` as a top level config item to use the platform signature. The request is fair.

Reply:
hmm maybe this is not possible because the `button` component doesn't have logic to handle this...

Reply:
I have removed this guard.